experimental/packages/opentelemetry-instrumentation/src/instrumentation.ts

```diff
@@ -173,4 +175,33 @@ export abstract class InstrumentationAbstract implements Instrumentation {
     | InstrumentationModuleDefinition
     | InstrumentationModuleDefinition[]
     | void;
+
+  /**
+   * Execute instrumentation event hook, if configured, and log any errors.
+   * Any semantics of the event type and eventInfo are defined by the specific instrumentation.
+   * @param hookHandler The optional hook handler which the user has configured via instrumentation config\
+   * @param eventName The name of the event being triggered, for logging purposes
+   * @param span The span to which the hook should be applied
+   * @param eventInfo The event info to be passed to the hook
+   */
+  protected runInstrumentationEventHook<EventInfoType>(
+    hookHandler: InstrumentationEventHook<EventInfoType> | undefined,
+    eventName: string,
+    span: Span,
```

Comment:
Are there hooks which are called before span has been started?

Reply:
There are at the moment in [instrumentation http](https://github.com/open-telemetry/opentelemetry-js/blob/5c8497152888c0a1886f4bc54af98ef1d9738589/experimental/packages/opentelemetry-instrumentation-http/src/types.ts#L75):

```js
export interface StartIncomingSpanCustomAttributeFunction {
  (request: IncomingMessage): SpanAttributes;
}

export interface StartOutgoingSpanCustomAttributeFunction {
  (request: RequestOptions): SpanAttributes;
}
```

I wonder if it is there to allow adding span attributes for sampling? 🤔

The common pattern at the moment is for manipulating a span after it is created, and for then there will be a benefit for not repeating the same invocation function in dozens of places.